### PR TITLE
updated to support more formatting options

### PR DESCRIPTION
**Pull request by fearthecowboy**
adds formatting options in modelerfour/prenamer so that the names are normalized better.

configuration added:

``` yaml
  # customization of the identifier normalization and naming provided by the prenamer.
  # pascalcase - MultiWordIdentifier 
  # camelcase - multiWordIdentifier 
  # snakecase - multi_word_identifier
  # uppercase - MULTI_WORD_IDENTIFIER 
  # kebabcase - multi-word-identifier 
  # default is the first one in the list below:
modelerfour:
  naming: 
    parameter: camelcase|pascalcase|snakecase|uppercase|kebabcase
    property: camelcase|pascalcase|snakecase|uppercase|kebabcase
    operation: pascalcase|camelcase|snakecase|uppercase|kebabcase
    operationGroup:  pascalcase|camelcase|snakecase|uppercase|kebabcase
    choice:  pascalcase|camelcase|snakecase|uppercase|kebabcase
    choiceValue:  pascalcase|camelcase|snakecase|uppercase|kebabcase
    constant:  pascalcase|camelcase|snakecase|uppercase|kebabcase
    type:  pascalcase|camelcase|snakecase|uppercase|kebabcase
```

### CODE REVIEW

File: modelerfour/prenamer/prenamer.ts [hunk 1]
```diff
@@ -17,172 +17,187 @@ function getNameOptions(typeName: string, components: Array<string>) {
...
+  async init() {
+    // get our configuration for this run.
+    this.options = await this.session.getValue('modelerfour', {});
+    const naming = this.options.naming || {};
+    this.format = {
+      parameter: formatStyle(naming.parameter, camelCase),
+      property: formatStyle(naming.property, camelCase),
+      operation: formatStyle(naming.operation, pascalCase),
+      operationGroup: formatStyle(naming.operationGroup, pascalCase),
+      choice: formatStyle(naming.schema, pascalCase),
```
Comment (daviwil) [inline modelerfour/prenamer/prenamer.ts:57]: Should this be `naming.choice`?

File: modelerfour/prenamer/prenamer.ts [hunk 2]
```diff
@@ -192,11 +207,12 @@ export class PreNamer {
 
       // handle operation parameters
       for (const parameter of collisions) {
+        console.log(`parameter collison fix on ${parameter.language.default.name} `)
```
Comment (daviwil) [inline modelerfour/prenamer/prenamer.ts:210]: Does the log need to be removed?

File: modelerfour/prenamer/prenamer.ts [hunk 3]
```diff
@@ -213,18 +229,18 @@ export class PreNamer {
...
 
     const usedNames = new Set(inlined.keys());
     for (const each of flattened /*.sort((a, b) => length(a.nameOptions) - length(b.nameOptions)) */) {
-      const ct = inlined.get(camelCase(each.language.default.camelName));
+      const ct = inlined.get(this.format.property(each.language.default.name));
       if (ct && ct > 1) {
-        // console.error(`Fixing collision on name ${each.name} #${ct} `);
-        const options = getNameOptions(each.schema.language.default.camelName, [each.language.default.camelName, ...values(each.flattenedNames)]);
-        each.language.default.camelName = camelCase(selectName(options, usedNames));
+        console.error(`Fixing collision on name ${each.language.default.name} #${ct} `);
```
Comment (daviwil) [inline modelerfour/prenamer/prenamer.ts:241]: Comment it back out or remove?

File: modelerfour/test/scenarios/a-playground/namer.yaml [hunk 4]
```diff
@@ -60,13 +58,13 @@ schemas: !<!Schemas>
...
             default:
-              name: YELLOW
+              name: Yellow
               description: ''
         - !<!ChoiceValue> 
           value: blacK
           language:
             default:
-              name: blacK
+              name: BlacK
```
Comment (daviwil) [inline modelerfour/test/scenarios/a-playground/namer.yaml:67]: Interesting that this still has a capitalized 'K' at the end when the `YELLOW` got cased correctly?

File: modelerfour/test/scenarios/a-playground/namer.yaml [hunk 5]
```diff
@@ -199,35 +192,33 @@ schemas: !<!Schemas>
...
-          originalName: field1
       protocol: !<!Protocols> {}
     - !<!Property> &ref_95
       schema: !<!NumberSchema> &ref_49
         type: integer
         precision: 64
         language: !<!Languages> 
           default:
-            name: typeForfield2
+            name: TypeForfield2
```
Comment (daviwil) [inline modelerfour/test/scenarios/a-playground/namer.yaml:210]: Should all these renamed entities have the `originalName` field added with the unnormalized name?

**Review by daviwil — APPROVED** [review]
Looks good overall but there are a few things that might need to be addressed before merging

File: modelerfour/prenamer/prenamer.ts (hunk 1 above)
Comment (fearthecowboy) [reply modelerfour/prenamer/prenamer.ts:57]: yeah

File: modelerfour/test/scenarios/a-playground/namer.yaml (hunk 4 above)
Comment (fearthecowboy) [reply modelerfour/test/scenarios/a-playground/namer.yaml:67]: It's two words ...

File: modelerfour/test/scenarios/a-playground/namer.yaml (hunk 5 above)
Comment (fearthecowboy) [reply modelerfour/test/scenarios/a-playground/namer.yaml:210]: I don't know that there's a lot of value left with the original name. if a generator needs it I could.

File: modelerfour/prenamer/prenamer.ts (hunk 3 above)
Comment (fearthecowboy) [reply modelerfour/prenamer/prenamer.ts:241]: fixed.

File: modelerfour/prenamer/prenamer.ts (hunk 2 above)
Comment (fearthecowboy) [reply modelerfour/prenamer/prenamer.ts:210]: fixed

File: modelerfour/prenamer/prenamer.ts (hunk 1 above)
Comment (fearthecowboy) [reply modelerfour/prenamer/prenamer.ts:57]: fixed

**Pull request merged**